conda/misc.py

```diff
@@ -177,6 +158,50 @@ def explicit(
     txn.execute()
 
 
+def _get_package_record_from_specs(specs: list[str]) -> Iterable[PackageCacheRecord]:
+    """Given a list of specs, find the corresponding PackageCacheRecord. If
+    some PackageCacheRecords are missing, raise an error.
+    """
+    specs_pcrecs = tuple(
+        [spec, next(PackageCacheData.query_all(spec), None)] for spec in specs
+    )
+
+    # Assert that every spec has a PackageCacheRecord
+    specs_with_missing_pcrecs = [
+        str(spec) for spec, pcrec in specs_pcrecs if pcrec is None
+    ]
+    if specs_with_missing_pcrecs:
+        if len(specs_with_missing_pcrecs) == len(specs_pcrecs):
+            raise AssertionError("No package cache records found")
+        else:
+            missing_precs_list = ", ".join(specs_with_missing_pcrecs)
+            raise AssertionError(
+                f"Missing package cache records for: {missing_precs_list}"
+            )
+    return [rec[1] for rec in specs_pcrecs]
+
+
+def get_package_records_from_explicit(lines: list[str]) -> Iterable[PackageCacheRecord]:
+    """Given the lines from an explicit.txt, create the PackageRecords for each of the
+    specified packages. This may require downloading the package, if it does not already
+    exist in the package cache.
+    """
+    # Extract the list of specs
+    fetch_specs = list(_match_specs_from_explicit(lines))
+
+    if context.dry_run:
+        raise DryRunExit()
+
+    # Try to get the specs from cache first
+    try:
+        return _get_package_record_from_specs(fetch_specs)
+    except AssertionError:
+        # If not found in cache, fetch them from the network and try again
+        pfe = ProgressiveFetchExtract(fetch_specs)
+        pfe.execute()
```

Comment:
This is a no-op if the packages are already cached, so maybe we don't need the `try / except`? It's true that it prints some text to stdout even if no actions are taken, but that can be fixed in a separate PR.